node/actors/bft/src/testonly/ut_harness.rs

```diff
@@ -0,0 +1,353 @@
+
+        self.consensus
+            .inner
+            .secret_key
+            .sign_msg(ConsensusMsg::LeaderPrepare(msg))
+    }
+
+    pub(crate) fn new_current_replica_commit(
+        &self,
+        mutate_fn: impl FnOnce(&mut ReplicaCommit),
+    ) -> Signed<ConsensusMsg> {
+        let mut msg = ReplicaCommit {
+            protocol_version: validator::CURRENT_VERSION,
+            view: self.consensus.replica.view,
+            proposal: self.consensus.replica.high_qc.message.proposal,
+        };
+
+        mutate_fn(&mut msg);
+
+        self.consensus
+            .inner
+            .secret_key
+            .sign_msg(ConsensusMsg::ReplicaCommit(msg))
+    }
+
+    pub(crate) fn new_rnd_leader_commit(
+        &mut self,
+        mutate_fn: impl FnOnce(&mut LeaderCommit),
+    ) -> Signed<ConsensusMsg> {
+        let mut msg = LeaderCommit {
+            protocol_version: validator::CURRENT_VERSION,
+            justification: self.rng().gen(),
+        };
+
+        mutate_fn(&mut msg);
+
+        self.consensus
+            .inner
+            .secret_key
+            .sign_msg(ConsensusMsg::LeaderCommit(msg))
+    }
+
+    pub(crate) async fn new_procedural_leader_prepare(&mut self) -> Signed<ConsensusMsg> {
+        let replica_prepare = self.new_current_replica_prepare(|_| {});
+        self.dispatch_replica_prepare(replica_prepare.clone())
+            .unwrap();
+        self.recv_signed().await.unwrap()
+    }
+
+    pub(crate) async fn new_procedural_replica_commit(&mut self) -> Signed<ConsensusMsg> {
+        let replica_prepare = self.new_current_replica_prepare(|_| {});
+        self.dispatch_replica_prepare(replica_prepare.clone())
+            .unwrap();
+        let leader_prepare = self.recv_signed().await.unwrap();
+        self.dispatch_leader_prepare(leader_prepare).await.unwrap();
+        self.recv_signed().await.unwrap()
+    }
+
+    pub(crate) async fn new_procedural_leader_commit(&mut self) -> Signed<ConsensusMsg> {
+        let replica_prepare = self.new_current_replica_prepare(|_| {});
+        self.dispatch_replica_prepare(replica_prepare.clone())
+            .unwrap();
+        let leader_prepare = self.recv_signed().await.unwrap();
+        self.dispatch_leader_prepare(leader_prepare).await.unwrap();
+        let replica_commit = self.recv_signed().await.unwrap();
+        self.dispatch_replica_commit(replica_commit).unwrap();
+        self.recv_signed().await.unwrap()
+    }
+
+    #[allow(clippy::result_large_err)]
+    pub(crate) fn dispatch_replica_prepare_many(
+        &mut self,
+        messages: Vec<ReplicaPrepare>,
+        keys: Vec<SecretKey>,
+    ) -> Result<(), ReplicaPrepareError> {
+        messages
+            .iter()
+            .enumerate()
+            .map(|(i, msg)| {
+                let signed = keys[i].sign_msg(ConsensusMsg::ReplicaPrepare(msg.clone()));
+                self.dispatch_replica_prepare(signed)
+            })
+            .last()
+            .unwrap()
```

Comment:
Are we indeed only interested in the last `Result` (as opposed to the cumulative one)?

Reply:
Only the last one is expected to succeed. In the new version I added assertions for the intermediate ones expected error.